### PR TITLE
Use hash_t internally in iterators and request handlers, wherever possible

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**

### CODE REVIEW

**Comment by gavin-norman-sociomantic** [comment]
D2 fixes.

**Comment by gavin-norman-sociomantic** [comment]
Rebased on master (post-v1.12.7 merge), added unittest. (Look! I actually paid attention to the coverage test!)

File: src/dhtnode/storage/StorageEngine.d [hunk 1]
```diff
@@ -192,6 +192,21 @@ public class StorageEngine : IStorageEngine
...
+
+        Ditto
+
+    ***************************************************************************/
+
+    public typeof(this) put ( hash_t key, cstring value,
+        bool trigger_listeners = true )
+    {
+        char[hash_t.sizeof * 2] key_str;
+        Hash.toHexString(key, key_str);
```
Comment (nemanja-boric-sociomantic) [inline src/dhtnode/storage/StorageEngine.d]: Hm, so the storage engine internally uses `hash_t`, but then you here convert the entire thing to `cstring` and then you call `this.put` which will convert it back to `hash_t`. Wouldn't it be better if we would move the implementation here and provide `cstring` methods as adapters to `hash_t`?

Comment (nemanja-boric-sociomantic) [reply src/dhtnode/storage/StorageEngine.d]: Sigh... This is already done, just the order of commits is messed up.

Comment (nemanja-boric-sociomantic) [reply src/dhtnode/storage/StorageEngine.d]: (in GitHub)

File: src/dhtnode/storage/StorageEngineStepIterator.d [hunk 2]
```diff
@@ -47,14 +48,21 @@ public class StorageEngineStepIterator
 
     ***************************************************************************/
 
-    private bool started;
+    private enum State
```
Comment (nemanja-boric-sociomantic) [inline src/dhtnode/storage/StorageEngineStepIterator.d]: Documentation is outdated.

File: src/dhtnode/storage/StorageEngineStepIterator.d [hunk 3]
```diff
@@ -47,14 +48,21 @@ public class StorageEngineStepIterator
...
+    {
+        Init,
+        Started,
+        Finished
+    }
+
+    private State state;
```
Comment (nemanja-boric-sociomantic) [inline src/dhtnode/storage/StorageEngineStepIterator.d]: ditto

File: src/dhtnode/storage/StorageEngineStepIterator.d [hunk 4]
```diff
@@ -148,14 +176,25 @@ public class StorageEngineStepIterator
...
+                    this.current_key);
+                break;
+            case Finished:
+                break;
+            default:
+                verify(false);
         }
+
+        if ( !more )
+            this.state = State.Finished;
```
Comment (nemanja-boric-sociomantic) [inline src/dhtnode/storage/StorageEngineStepIterator.d]: This thing with the `State` and refactoring here could be a separate commit, right?

Comment (gavin-norman-sociomantic) [reply src/dhtnode/storage/StorageEngineStepIterator.d]: Hmm... yeah, that commit is a bit weird. Just trying to figure out what's happening there :/

Comment (gavin-norman-sociomantic) [reply src/dhtnode/storage/StorageEngineStepIterator.d]: Added a bit of explanation in the commit message.
